server/app/Icepeak/Main.hs

```diff
@@ -79,31 +79,52 @@ runCore core = do
 
   -- start threads
   pops <- Async.async $ Core.runCommandLoop core
-  sync <- Async.async $ Core.runSyncTimer core
   upds <- Async.async $ WebsocketServer.processUpdates core
   serv <- Async.async $ Server.runServer logger wsServer httpServer
+  sync <- Async.async (Core.runSyncTimer core)
   metrics <- Async.async
     $ forM_ (configMetricsEndpoint config) (MetricsServer.runMetricsServer logger)
+
   installHandlers core serv
   logAuthSettings config logger
   logQueueSettings config logger
   logSyncSettings config logger
   postLog logger "System online. ** robot sounds **"
 
-  waitLog "core" logger pops
-  waitLog "web sockets server" logger upds
-  waitLog "http server" logger serv
-  -- kill auxiliary threads when the main threads ended
-  Async.cancel metrics
-  Async.cancel sync
-
--- | Wait for an Async computation to exit and log unexpected exceptions.
-waitLog :: Text -> Logger -> Async () -> IO ()
-waitLog name logger action = Async.waitCatch action >>= handleLog where
-  handleLog (Left exc)
-    | Just (_ :: AsyncException) <- fromException exc = pure ()
-    | otherwise = Logger.postLog logger $ name <> " stopped with an exception: " <> Text.pack (show exc)
-  handleLog (Right _) = pure ()
+  runCoreResult <- waitFirstRunCoreResult pops upds serv
+  logRunCoreResult logger runCoreResult
+
+  -- kill all threads when one of the main threads ended
+  mapM_ Async.cancel [pops, upds, serv, metrics, sync]
```

Comment:
What happens when one of the threads already ended? Is `cancel` a no-op then?

The core should be stopped with `Core.postQuit`, as that will cause it to save the current state to disk before stopping. Although the came could be achieved by catching exceptions inside the handler loop in the core.

But one could also argue that in cases of unhandled exceptions the data might be corrupted, and therefore should not be saved.

Thinking about this: I wonder what exceptions can actually happen here anyway, besides failing to open the socket. Maybe we should be more explicit about those errors and more thoughtful about the order in which we start up things.

Reply:
`cancel` is indeed a no-op if the thread already ended. I've changed the code to use `Core.postQuit` and wait for the core to stop.